internal/v4/api_test.go

```diff
@@ -2455,6 +2450,102 @@ func (s *APISuite) TestHash256Laziness(c *gc.C) {
 	c.Assert(entity.BlobHash256, gc.Not(gc.Equals), "")
 }
 
+var urlChannelResolvingEntities = []struct {
+	id      *router.ResolvedURL
+	channel mongodoc.Channel
+}{{
+	id:      newResolvedURL("~charmers/precise/wordpress-0", 0),
+	channel: mongodoc.StableChannel,
+}, {
+	id:      newResolvedURL("~charmers/precise/wordpress-1", 1),
+	channel: mongodoc.DevelopmentChannel,
+}, {
+	id:      newResolvedURL("~charmers/precise/wordpress-2", 2),
+	channel: mongodoc.UnpublishedChannel,
+}, {
+	id:      newResolvedURL("~charmers/trusty/mysql-0", 0),
+	channel: mongodoc.UnpublishedChannel,
+}}
+
+var urlChannelResolvingTests = []struct {
+	url          string
+	channel      mongodoc.Channel
+	expectURL    string
+	expectStatus int
+	expectError  params.Error
+}{{
+	url:       "wordpress",
+	expectURL: "cs:precise/wordpress-0",
+}, {
+	url:       "wordpress",
+	channel:   mongodoc.StableChannel,
+	expectURL: "cs:precise/wordpress-0",
+}, {
+	url:       "wordpress",
+	channel:   mongodoc.DevelopmentChannel,
+	expectURL: "cs:precise/wordpress-1",
+}, {
+	url:       "wordpress",
+	channel:   mongodoc.UnpublishedChannel,
+	expectURL: "cs:precise/wordpress-2",
+}, {
+	url:       "~charmers/precise/wordpress",
+	channel:   mongodoc.StableChannel,
+	expectURL: "cs:~charmers/precise/wordpress-0",
+}, {
+	url:          "mysql",
+	expectStatus: http.StatusNotFound,
+	expectError: params.Error{
+		Message: `no matching charm or bundle for "cs:mysql"`,
+		Code:    params.ErrNotFound,
+	},
+}, {
+	url:          "mysql",
+	channel:      "unknown",
+	expectStatus: http.StatusBadRequest,
+	expectError: params.Error{
+		Message: `invalid channel "unknown" specified in request`,
+		Code:    params.ErrBadRequest,
+	},
+}}
+
+func (s *APISuite) TestURLChannelResolving(c *gc.C) {
+	s.discharge = dischargeForUser("charmers")
+	for _, add := range urlChannelResolvingEntities {
+		err := s.store.AddCharmWithArchive(add.id, storetesting.NewCharm(nil))
+		c.Assert(err, gc.IsNil)
+		if add.channel != mongodoc.UnpublishedChannel {
+			err = s.store.Publish(add.id, add.channel)
+			c.Assert(err, gc.IsNil)
+		}
+	}
+	for i, test := range urlChannelResolvingTests {
```

Comment:
I don't think so. We're using meta/any here only because it's the easiest way to tell us what the resolved id was. Given that the implementation only resolved URLs in one place, I don't think we'd be testing anything else by adding tests for other endpoints (there are lots of endpoints that return entity ids).